### DISCUSSION TITLE
createSharedComposable vs createGlobalState

### DISCUSSION THREAD

**Discussion by aphex** [vueuse/vueuse#2845] [Q&A]
Can anyone help me better understand why you would use one of these over the other? They both seem to accomplish similar goals of sharing data across vue instances.

**Comment by darrylnoakes** [comment]
The docs say "multiple Vue instances," but it is multiple _component_ instances.

`createGlobalState` allows state to be shared across component instances. Like a minimal version of Pinia.

`createSharedComposable` allows the internal state of a single composable to be reused across multiple uses and components instances.
So, in the `useMouse` example in the docs, only one set of event listeners will be attached, and all the components that use the shared composable can reuse those same listeners.

**Reply by aphex** [reply]
Hmmm not sure I am totally seeing the difference here. If you added a `window` listener for mouseMove in `createGlobalState` wouldnt it also only ever be one set of event listeners in all components that share that state? The composable function only seems to ever run once. 

Looking at the code for `createGlobalState` here
https://github.com/vueuse/vueuse/blob/main/packages/shared/createGlobalState/index.ts
![image](https://user-images.githubusercontent.com/152484/223283732-c2c3cb58-e617-45d2-89ef-47c503c1325a.png)

Comparing it to the code here for `createSharedComposable`
https://github.com/vueuse/vueuse/blob/main/packages/shared/createSharedComposable/index.ts
![image](https://user-images.githubusercontent.com/152484/223283746-9bedd6a8-d4d9-4bd5-b7f5-42b7893cf357.png)

These seem very similar in practice. The both isolate the execution of the wrapped composable function, however the main difference I see is that `createSharedComposable` will track subscriptions. Ticking them up and down as different components "use" the shared composable. When subscribers are exhausted it will clean up, and then re-run the wrapped composable function when needed. 

Where-as `createGlobalState` seems to only run the wrapped composable function once, and will never dispose it. This seems like it would be problematic for hot module reload. 

Maybe what it comes down to is `createGlobalState` is good for composables that have no listeners or scope side effects to dispose ever? Something you only ever want to have run one time, and never again. Where `createSharedComposable` is good when you want to add/remove side effects depending on if the composable is actively in use "right now"?

Just not sure why one wouldn't always want that tracking and adding/removing of listeners and side effects.

**Reply by darrylnoakes** [reply]
`createGlobalState` is intended to share _state_. The function it takes is like the Options API `data()` function; it is simply for creating the shared data.

`createSharedComposable` takes an existing _composable_ and creates a _shared version of it_, which shares a single "instance" of the original composable across uses of the shared version.
It will handle cleaning up effects, etc.

**Reply by aphex** [reply]
Ok that's not to bad, `createGlobalState` is meant for data and `createShartedComposable` create a _shared composable_.

I truly don't mean to be difficult haha, I am genuinely struggling trying to put these things in categories to make a choice. With that would you also agree that `createSharedComposable` **could** be used for sharing state, if the composable you wrap is just creating reactive data and returning it? It does seem that `shared` is capable of doing everything `state` does, but maybe just has some overhead?

Maybe its best to look at it like you should try to start with `createGlobalState` as it is lighter weight, doesn't have the overhead of tracking or scope disposal. Then you move up to `createSharedState` if your data composable gains scope side effects that could benefit from cleanup when not in use?

thanks again for the conversation

**Reply by darrylnoakes** [reply]
> With that would you also agree that `createSharedComposable` **could** be used for sharing state, if the composable you wrap is just creating reactive data and returning it? It does seem that `shared` is capable of doing everything `state` does, but maybe just has some overhead?

Well, with `createGlobalState`, the global state is maintained across uses of the created composable, _even if there is nothing currently using the state_. For example a component instance could be mounted, use the state, and then be destroyed. When another component instance uses the created composable, _the state from the previous use will have been maintained_.

With `createSharedComposable`, the number of "subscribers" is tracked. _If there are no longer any effect scopes using the state, the state is destroyed_. This obviously wouldn't work as a state store, as the state would be lost if no component currently had a handle on it.
However, it allows for efficient use of things like event listeners, as in the `useMouse` example.

**Reply by darrylnoakes** [reply]
In short:

`createGlobalState` allows state to be shared across components, _with the state primarily controlled by the components_.

`createSharedComposable` allows state to be shared across components and resources (such as event listeners) reused, _either with the state primarily controlled by the composable itself_ (updates in response to event listeners, etc.), _or in some kind of command/lock system, such as shared and safe access to storage_ (for example, the storage would only need to be connected to by the single instance of the composable, and multiple components could safely access it via the composable).

**Reply by aphex** [reply]
Oh man, that's perfect and so obvious now that you have said it. Thanks again for talking through it, totally just clicked.
